File: uikit/src/main/java/com/sendbird/uikit/widgets/SuggestedMentionPreview.java
package com.sendbird.uikit.widgets;

import android.content.Context;
import android.content.res.TypedArray;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.sendbird.android.user.User;
import com.sendbird.uikit.R;
import com.sendbird.uikit.databinding.SbViewSuggestedMentionListItemBinding;
import com.sendbird.uikit.utils.UserUtils;
import com.sendbird.uikit.utils.ViewUtils;

public class SuggestedMentionPreview extends FrameLayout {
    private final SbViewSuggestedMentionListItemBinding binding;
    private final int nicknameTextAppearance;
    private final int emptyNicknameTextAppearance;

    public SuggestedMentionPreview(@NonNull Context context) {
        this(context, null);
    }

    public SuggestedMentionPreview(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, R.attr.sb_widget_suggested_mention_preview);
    }

    public SuggestedMentionPreview(@NonNull Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.UserPreview, defStyle, 0);
        try {
            this.binding = SbViewSuggestedMentionListItemBinding.inflate(LayoutInflater.from(getContext()));
            addView(binding.getRoot(), ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);

            int background = a.getResourceId(R.styleable.UserPreview_sb_member_preview_background, R.drawable.selector_rectangle_light);
            int descAppearance = a.getResourceId(R.styleable.UserPreview_sb_member_preview_description_appearance, R.style.SendbirdBody2OnLight02);
            this.nicknameTextAppearance = a.getResourceId(R.styleable.UserPreview_sb_member_preview_nickname_appearance, R.style.SendbirdSubtitle2OnLight01);
            this.emptyNicknameTextAppearance = a.getResourceId(R.styleable.UserPreview_sb_mention_empty_nickname_appearance, R.style.SendbirdBody2OnLight03);
            binding.getRoot().setBackgroundResource(background);
            binding.tvNickname.setEllipsize(TextUtils.TruncateAt.END);
            binding.tvNickname.setMaxLines(1);
            binding.tvDescription.setTextAppearance(context, descAppearance);
        } finally {
            a.recycle();
        }
    }

    @NonNull
    public View getLayout() {
        return this;
    }

    @NonNull
    public SbViewSuggestedMentionListItemBinding getBinding() {
        return binding;
    }

    @Override
    public void setOnClickListener(@Nullable OnClickListener listener) {
        binding.vgMemberItem.setOnClickListener(listener);
    }

    @Override
    public void setOnLongClickListener(@Nullable OnLongClickListener listener) {
        binding.vgMemberItem.setOnLongClickListener(listener);
    }

    public void setOnProfileClickListener(@Nullable OnClickListener listener) {
        binding.ivProfile.setOnClickListener(listener);
    }

    public void setDescription(@Nullable CharSequence text) {
        binding.tvDescription.setText(text);
    }

    public void setName(@Nullable CharSequence name) {
        binding.tvNickname.setText(name);
    }

    public void setImageFromUrl(@Nullable String url) {
        ViewUtils.drawProfile(binding.ivProfile, url);
    }

    public void drawUser(@NonNull User user, boolean showUserId) {
        final Context context = getContext();
        final String nickname = UserUtils.getDisplayName(getContext(), user);

        if (TextUtils.isEmpty(user.getNickname())) {
            binding.tvNickname.setTextAppearance(context, emptyNicknameTextAppearance);
        } else {
            binding.tvNickname.setTextAppearance(context, nicknameTextAppearance);
        }
        setName(nickname);

        if (showUserId) {
            final String description = user.getUserId();
            setDescription(description);
        }
        setImageFromUrl(user.getProfileUrl());
    }
}

File: uikit/src/main/java/com/sendbird/uikit/internal/ui/widgets/SuggestedMentionPreview.kt
package com.sendbird.uikit.internal.ui.widgets

import android.content.Context
import android.text.TextUtils
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import com.sendbird.android.user.User
import com.sendbird.uikit.R
import com.sendbird.uikit.databinding.SbViewSuggestedMentionListItemBinding
import com.sendbird.uikit.internal.extensions.setAppearance
import com.sendbird.uikit.utils.UserUtils
import com.sendbird.uikit.utils.ViewUtils

internal class SuggestedMentionPreview @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = R.attr.sb_widget_suggested_mention_preview
) : FrameLayout(context, attrs, defStyle) {
    val binding: SbViewSuggestedMentionListItemBinding
    val layout: View
        get() = this
    private var nicknameTextAppearance: Int
    private var emptyNicknameTextAppearance: Int

    override fun setOnClickListener(listener: OnClickListener?) {
        binding.vgMemberItem.setOnClickListener(listener)
    }

    override fun setOnLongClickListener(listener: OnLongClickListener?) {
        binding.vgMemberItem.setOnLongClickListener(listener)
    }

    fun setOnProfileClickListener(listener: OnClickListener?) {
        binding.ivProfile.setOnClickListener(listener)
    }

    fun setDescription(text: CharSequence?) {
        binding.tvDescription.text = text
    }

    fun setName(name: CharSequence?) {
        binding.tvNickname.text = name
    }

    private fun setImageFromUrl(url: String?, plainUrl: String?) {
        ViewUtils.drawProfile(binding.ivProfile, url, plainUrl)
    }

    fun drawUser(user: User, showUserId: Boolean) {
        val context = context
        val nickname = UserUtils.getDisplayName(getContext(), user)
        if (TextUtils.isEmpty(user.nickname)) {
            binding.tvNickname.setAppearance(context, emptyNicknameTextAppearance)
        } else {
            binding.tvNickname.setAppearance(context, nicknameTextAppearance)
        }
        setName(nickname)
        if (showUserId) {
            val description = user.userId
            setDescription(description)
        }
        setImageFromUrl(user.profileUrl, user.plainProfileImageUrl)
    }

    init {
        val a = context.theme.obtainStyledAttributes(attrs, R.styleable.UserPreview, defStyle, 0)
        try {
            binding = SbViewSuggestedMentionListItemBinding.inflate(LayoutInflater.from(getContext()))
            addView(binding.root, ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
            val background = a.getResourceId(
                R.styleable.UserPreview_sb_member_preview_background,
                R.drawable.selector_rectangle_light
            )
            val descAppearance = a.getResourceId(
                R.styleable.UserPreview_sb_member_preview_description_appearance,
                R.style.SendbirdBody2OnLight02
            )
            nicknameTextAppearance = a.getResourceId(
                R.styleable.UserPreview_sb_member_preview_nickname_appearance,
                R.style.SendbirdSubtitle2OnLight01
            )
            emptyNicknameTextAppearance = a.getResourceId(
                R.styleable.UserPreview_sb_mention_empty_nickname_appearance,
                R.style.SendbirdBody2OnLight03
            )
            binding.root.setBackgroundResource(background)
            binding.tvNickname.ellipsize = TextUtils.TruncateAt.END
            binding.tvNickname.maxLines = 1
            binding.tvDescription.setAppearance(context, descAppearance)
        } finally {
            a.recycle()
        }
    }
}
